### PR TITLE
[Test] Verify monitoring stack image version using ceph orchestrator

### PR DESCRIPTION
**Pull request by sayaleeraut**
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by sayaleeraut** [comment]
Tox results -

py3: commands[1]> flake8
py3: commands[2]> isort -c .
Skipped 3 files
py3: commands[3]> black --check --diff .
All done! ✨ 🍰 ✨
1046 files would be left unchanged.
py3: commands[4]> yamllint -d relaxed --no-warnings .
.pkg: _exit> python /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
  py3: OK (23.69=setup[3.80]+cmd[0.49,1.41,1.28,2.69,14.03] seconds)
  congratulations :) (23.72 seconds)

File: tests/cephadm/test_ceph_orch_for_monitoring_image_version.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    cephadm = CephAdm(installer_node)
+    monitoring_images = config.get("monitoring_images")
+
+    # Get the version using podman command
+    prometheus_img = monitoring_images["ose-prometheus"]
+    alertmanager_img = monitoring_images["ose-prometheus-alertmanager"]
+    node_exporter_img = monitoring_images["ose-prometheus-node-exporter"]
+
+    # Get the version using podman command
+    cmd = f'podman run -it --rm --user root --entrypoint /bin/bash {prometheus_img} -c "prometheus --version"'
```
Comment (pranavprakash20) [inline tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:24]: maybe we can loop these operations?
something like

base_cmd = 'podman run -it --rm --user root --entrypoint /bin/bash {} -c "{} --version"'

for img, mon_stack in ([prometheus_img, 'prometheus'],
                       [alertmanager_img, "alertmanager"],
                       [node_exporter_img, "node_exporter"]):
    cmd = base_cmd.format(img, mon_stack)
    out, _ = installer_node.exec_command(sudo=True, cmd=cmd)
    _version = re.search(r"version\s+(\d+\.\d+\.\d+)", out)
    
    
    Just a suggestion. IF this can't be done or will cause logic problem with existing approach, please ignore.

Comment (vamahaja) [inline tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:24]: Instead of podman command use containers library - https://github.com/red-hat-storage/cephci/blob/master/cli/utilities/containers.py

File: tests/cephadm/test_ceph_orch_for_monitoring_image_version.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    out, _ = installer_node.exec_command(sudo=True, cmd=cmd)
+    _version = re.search(r"version\s+(\d+\.\d+\.\d+)", out)
```
Comment (vamahaja) [inline tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:26]: For getting version consider image type as well as build type RH & IBM.

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/cephadm/test_ceph_orch_for_monitoring_image_version.py (hunk 1 above)
Comment (sayaleeraut) [reply tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:24]: Hey @pranavprakash20  I need to store the output of  re.search(r"version\s+(\d+.\d+.\d+)", out) in a variable which I can use later for validation. 
Maybe I can add somethingv as below-
```services = {
    "prometheus": "_prom_vers",
    "alertmanager": "_alert_vers",
    "node_exporter": "_node_vers"
}

for service, version_variable in services.items():
    image = monitoring_images.get(service)
    if image:
        cmd = f"podman run -it --rm --user root --entrypoint /bin/bash {image} -c \"{service} --version\""
        out, _ = installer_node.exec_command(sudo=True, cmd=cmd)
        _version = re.search(r"version\s+(\d+\.\d+\.\d+)", out)
        if _version:
            globals()[version_variable] = _version.group(1)
```
   Let me try and update the PR accordingly.

Comment (sayaleeraut) [reply tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:24]: Hey @pranavprakash20  Its complicating the code actually, because we have images also to iterate through, so I will prefer to leave it as it is, let me know if you agree.

File: tests/cephadm/test_ceph_orch_for_monitoring_image_version.py (hunk 2 above)
Comment (sayaleeraut) [reply tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:42]: Done

File: tests/cephadm/test_ceph_orch_for_monitoring_image_version.py (hunk 1 above)
Comment (sayaleeraut) [reply tests/cephadm/test_ceph_orch_for_monitoring_image_version.py]: Done

File: cli/utilities/containers.py [hunk 3]
```diff
@@ -76,6 +80,11 @@ def run(
             detach_key (list): List of detach operation
             detach (str): Image name to detach
             cmds (str): Other commands to be executed
+            interactive (bool): Keep stdin open even if not attached
+            tty (bool): Allocate a pseudo-TTY
+            user (str): User to run container as
+            entry_point (str): Entry point command or executable to be run
+            inline_cmd (str): Inline command to be executed withing container
```
Comment (vamahaja) [inline cli/utilities/containers.py:87]: `inline_cmd`???

File: tests/cephadm/test_ceph_orch_for_monitoring_image_version.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    for monitoring stack images
+    """
+    config = kw["config"]
+    installer_node = ceph_cluster.get_nodes(role="installer")[0]
+    cephadm = CephAdm(installer_node)
+    monitoring_images = config.get("monitoring_images")
+
+    # Get the build type
+    build_type = "ibm" if config.get("ibm_build") else "rh"
+    if build_type == "rh":
```
Comment (vamahaja) [inline tests/cephadm/test_ceph_orch_for_monitoring_image_version.py:21]: We have separate suites for IBM & RH builds. No need for this check.

File: suites/reef/cephadm/tier1-container-cli-args.yaml [hunk 5]
```diff
@@ -285,3 +286,19 @@ tests:
...
       abort-on-fail: true
+
+  - test:
+      name: Validate ceph orch displays image version
+      desc: Test the monitoring stack image version using ceph orch
+      module: test_ceph_orch_for_monitoring_image_version.py
+      polarion-id: CEPH-83589764
+      config:
+        monitoring_images:
+          rhcs:
```
Comment (vamahaja) [inline suites/reef/cephadm/tier1-container-cli-args.yaml:297]: Create separate test suite for ibm build

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Review by pranavprakash20 — DISMISSED** [review]
LGTM

**Review by pranavprakash20 — APPROVED** [review]
LGTM

**Review by MohitBis — DISMISSED** [review]
LGTM

File: cli/utilities/containers.py [hunk 6]
```diff
@@ -113,17 +121,33 @@ def run(
...
+            cmd += f" --entrypoint {entry_point}"
 
         if image:
             cmd += f" {image}"
 
         if cmds:
             cmd += f" {cmds}"
 
-        if self.execute(sudo=True, long_running=True, cmd=cmd):
+        out = self.execute(sudo=True, cmd=cmd)
```
Comment (vamahaja) [inline cli/utilities/containers.py:145]: Revert this change, this should be long running command. It will affect existing test cases.

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/utilities/containers.py [hunk 7]
```diff
@@ -113,17 +121,33 @@ def run(
...
             cmd += f" {image}"
 
         if cmds:
             cmd += f" {cmds}"
 
-        if self.execute(sudo=True, long_running=True, cmd=cmd):
+        out = self.execute(sudo=True, long_running=True, cmd=cmd)
+        if not out:
             raise ContainerRegistryError(f"Failed to run {cmd} on container")
+        if isinstance(out, tuple):
```
Comment (vamahaja) [inline cli/utilities/containers.py:148]: This is incorrect, when `long_running` set to `True`, it won't return tuple

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/utilities/containers.py (hunk 7 above)
Comment (sayaleeraut) [reply cli/utilities/containers.py:149]: Made the correction.

**Review by MohitBis — APPROVED** [review]
LGTM

**Comment by vamahaja** [comment]
Kindly share test execution logs

**Review by tintumathew10 — COMMENTED** [review]
LGTM

**Comment by sayaleeraut** [comment]
Test execution logs - http://magna002.ceph.redhat.com/cephci-jenkins/saraut/PR_3687/

**Review by vamahaja — APPROVED** [review]
